redsky/streams.py

```diff
@@ -0,0 +1,630 @@
+        Stream.__init__(self, children=children)
+
+    def update(self, x, who=None):
+        if self.missing and who in self.missing:
+            self.missing.remove(who)
+
+        self.last[self.children.index(who)] = x
+        if not self.missing:
+            tup = tuple(self.last)
+            if tup and hasattr(tup[0], '__stream_merge__'):
+                tup = tup[0].__stream_merge__(*tup[1:])
+            return self.emit(tup)
+
+
+class concat(Stream):
+    def update(self, x, who=None):
+        L = []
+        for item in x:
+            y = self.emit(item)
+            if type(y) is list:
+                L.extend(y)
+            else:
+                L.append(y)
+        return L
+
+
+class unique(Stream):
+    def __init__(self, child, history=None, key=identity):
+        self.seen = dict()
+        self.key = key
+        if history:
+            from zict import LRU
+            self.seen = LRU(history, self.seen)
+
+        Stream.__init__(self, child)
+
+    def update(self, x, who=None):
+        y = self.key(x)
+        if y not in self.seen:
+            self.seen[y] = 1
+            return self.emit(x)
+
+
+class union(Stream):
+    def update(self, x, who=None):
+        return self.emit(x)
+
+
+class collect(Stream):
+    def __init__(self, child, cache=None):
+        if cache is None:
+            cache = deque()
+        self.cache = cache
+
+        Stream.__init__(self, child)
+
+    def update(self, x, who=None):
+        self.cache.append(x)
+
+    def flush(self, _=None):
+        out = tuple(self.cache)
+        self.emit(out)
+        self.cache.clear()
+
+
+class starmap(Stream):
+    def __init__(self, func, child, raw=False, **kwargs):
+        self.func = func
+        self.kwargs = kwargs
+        self.raw = raw
+
+        Stream.__init__(self, child)
+
+    def update(self, x, who=None):
+        if not self.raw and hasattr(x, '__stream_map__'):
+            result = x.__stream_map__(self.func, **self.kwargs)
+        else:
+            result = self.func(*x, **self.kwargs)
+
+        return self.emit(result)
+
+
+class dstarmap(Stream):
+    def __init__(self, func, child, raw=False, output_info=None,
+                 input_info=None, **kwargs):
+        self.func = func
+        self.kwargs = kwargs
+        self.raw = raw
+
+        Stream.__init__(self, child, output_info=output_info,
+                        input_info=input_info, **kwargs)
+        ### TODO: update the provenence now that we have the func
+
+    def update(self, x, who=None):
+        # massage the pair(s)
+        res = self.dispatch(x)
+        # if we are giving back a new doc, just emit it
+        if isinstance(res, tuple) and res[0] in ['start', 'descriptor',
+                                                 'stop']:
+            return self.emit(res)
+        try:
+            # we need to expose the raw event data
+            res = self.event_guts(res)
+            if not self.raw and hasattr(x, '__stream_map__'):
```

Comment:
So to step back, I was suggesting either the incoming data be a new object (ex: `DataStream(doctuple)`) that we distinguish, or we use a sentinel (a property we can check for that we're certain not to find in another class). I think this suggestion still goes the sentinel route.
Have I misunderstood perhaps?

If yes, we need something to distinguish the document with a property we can ascertain no tuple that we create will have. The problem is the stream mapping is recursive, and so if the result of the function is a list/tuple whose first element is `'start'` etc... then it will wrongly be interpreted as a document tuple and dispatched to the wrong function. We could also check if the document is conformant to some schema but I feel that may be too much overhead. Maybe a stream document can be assumed to have some sentinel? Like `doc['databroker-version']` or something?